File: CSCI141/hw5/passwordChecker.py
# Author : Martin Smith
# Description : HW 5, CSCI 141
# Date : November 11th 2016
# Assignment : passwordChecker


# have function validatePassword
def validatePassword (userPassword) :
    import string

    #if there is a space in password, return False
    if " " in userPassword:
        return False

    #start with all values being False
    digits = False
    upperCase = False
    lowerCase = False

    #have a for loop to determine what characters are in the password
    for variable in (userPassword):
        if variable not in string.ascii_lowercase:
            if variable in string.digits:
                digits = True
            if variable in string.ascii_uppercase:
                upperCase = True
        elif variable not in string.ascii_uppercase:
            lowerCase = True

    #determine what value to return depending on what the for loop outputted
    if digits != True:
        return False
    elif upperCase != True:
        return False
    elif lowerCase != True:
        return False
    else:
        return True


# have main function
def main () :

    import string

    #have while loop
    inputtingPasswords = False

    #print explanatory statement
    print("Please input passwords one-by-one for validation.")
    print("Input END after you've entered your last password.")
    
    #have a value for the empty list
    passwordList = []
    
    #ask the user for passwords until they don't want to enter anymore
    while inputtingPasswords == False:
        passwordInput = input("")

        #add each password to the list
        passwordList.append(passwordInput)

        #when user enters "END", end the while loop
        if passwordInput == "END":
            inputtingPasswords = True

    #have a for loop to evaluate each password in the list
    for x in range(0, len(passwordList)-1):

        #evaluate if the validatePassword function returns True or False
        passwordRight = validatePassword (passwordList[x])

        #if the value returns True, tell user their password is good
        if passwordRight ==True:
            print("Password", passwordList[x], "is good.")
        else:

            #if there is a space in the password, tell user their password
            #is bad and explain why
            if " " in passwordList[x]:
                print("Password", passwordList[x],"is bad. It has a white space.")
            else:

                #have all values equal False until evaluated in for loop
                upperCase = False
                lowerCase = False
                digits = False

                #evaluate each character in each for loop to determine
                #if there is upper case, lower case, or numbers in the
                #said password
                for value in passwordList[x]:
                    if value not in string.digits:
                        if value not in string.ascii_lowercase:
                            upperCase = True
                        else:
                            if value not in string.ascii_uppercase:
                                lowerCase = True
                    else:
                        digits = True

                #based on the evaluataions, determine how to tell user why their password
                #is bad
                if digits == False:
                    print ("Password", passwordList[x], "is bad. It has no digits.")
                elif upperCase == False:
                    print ("Password", passwordList[x], "is bad. It has no upper case letters.")
                else:
                    if lowerCase == False:
                        print("Password", passwordList[x], "is bad. It has no lower casee letters.")

#declare main function
main()

File: PasswordChecker.java
// Author : Martin Smith
// Description : CSCI 145, Lab 2
// Date : April 13th 2017
// Assignment : PasswordChecker


//import the scanner utility before starting the program
import java.util.*;

public class PasswordChecker {
   
   
   // have the main function
   
   public static void main (String[]args){
      
      // have a while loop and a boolean value so user continues to input passwords until they type a valid one
      
      boolean validPassword = false;
      
      while (validPassword == false){
         
         //ask user for their password
         
         Scanner console = new Scanner(System.in);
         System.out.print("Please enter a password: ");
         String password = console.next();
         
         // have an if-else statement to determine if the user's password is a valid one
         // invoke the isValid function in order to do so

         //if the function returns false, print "Invalid Password!", else print "Valid Password!" and break while loop
         
         if (isValid(password) == false) {
            System.out.println("Invalid Password!");
         }else{
            System.out.println("Valid Password!");
            validPassword = true;
         }
      }
   }
   
   // have the boolean isValid function to evaluate the password based on its parameters,
   // and return the value as either true or false
   
   public static boolean isValid(String s){
      
      //declare the parameters needed to evaluate the password as false, so it can be later
      //referenced in the function
      boolean eightCharacters = false;
      boolean fourLetters = false;
      boolean twoDigits = false;
      boolean twoUppercase = false;
      boolean onlyLetterDigits = false;
   
      //to determine the number of letters, numbers, and uppercase letters in the password, have int countLetter,
      //int countDigit and int countUppercase equal 0
      int countLetter = 0;
      int countDigit = 0;
      int countUppercase = 0;
      
      //have a for loop that will evaluate each character in the password, and determine how many letters,      
      //digits, and uppercase letters there are in the password
      for (int start=0;start <= (s.length()-1);start++){
         //go through each character
         char value = s.charAt(start);
         
         //have an if statement to deterine if the said character is a letter
         if (Character.isLetter(value)){
            countLetter++;
         }
         if (Character.isDigit(value)){
            countDigit++;
         }
         if (Character.isUpperCase(value)){
            countUppercase++;
         }
      }
      
      //use if-else statements to evaluate if the password has satisified the parameters using the length() string method,
      //and the values outputted from the for loop
      if (s.length() >= 8) {
         eightCharacters = true;
      }
      
      if (countLetter >= 4) {
         fourLetters = true;
      }
      
      if (countDigit >= 2) {
         twoDigits = true;
      }
      
      if (countUppercase >= 2) {
         twoUppercase = true;
      }
      
      if (countLetter + countDigit == s.length()){
         onlyLetterDigits = true;
      }
      
      //if all boolean values are true, return true, else return false\
      if (eightCharacters == true && fourLetters == true && twoDigits == true && twoUppercase == true && onlyLetterDigits == true){
         return true;
      } else {
         return false;
      }   
   }
}